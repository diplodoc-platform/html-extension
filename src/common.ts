import type {HtmlController} from './runtime/HtmlController';

export const GLOBAL_SYMBOL: unique symbol = Symbol.for('diplodocHtml');

declare global {
    interface Window {
        [GLOBAL_SYMBOL]: HtmlController;
    }
}
