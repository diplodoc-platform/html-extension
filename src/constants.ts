export const GLOBAL_SYMBOL = Symbol.for('diplodocHtml');
export const QUEUE_SYMBOL = Symbol.for('queue');

export const DEFAULT_IFRAME_HEIGHT_PADDING = 34;

export const SRCDOC_TOKEN_TYPE = 'yfm_html_block';
export const SHADOW_TOKEN_TYPE = 'yfm_html_block_shadow';
export const ISOLATED_TOKEN_TYPE = 'yfm_html_block_isolated';

export const DATAATTR_SANDBOX_MODE = 'data-yfm-sandbox-mode';
export const DATAATTR_SANDBOX_CONTENT = 'data-yfm-sandbox-content';
export const DATAATTR_ISOLATED_SANDBOX_BASE_TARGET = 'data-yfm-sandbox-base-target';
export const DATAATTR_ISOLATED_SANDBOX_PREFERRED_HOST =
    'data-yfm-sandbox-preferred-isolated-host-uri';
