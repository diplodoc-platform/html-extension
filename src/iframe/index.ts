import {PostMessageChannel} from '../rpcAdapter/PostMessageChannel';
import {APIPublisher} from '../rpcAdapter/APIPublisher';

import {IFrameController} from './IFrameController';

export type {Commands, Events} from './IFrameController';

const $$PublisherInstanceSymbol = Symbol.for('$$RPCAPIPublisher');

const onDOMReady = () => {
    const controller = new IFrameController(globalThis.window);
    const publisher = new APIPublisher(new PostMessageChannel(globalThis.parent));

    publisher
        .onCommand('setStyles', controller.setStyles)
        .onCommand('setClassNames', controller.setClassNames)
        .onCommand('replaceHTML', controller.replaceHTML)
        .onCommand('setBaseTarget', controller.setBaseTarget);

    controller.on('resize', (rect) => publisher.dispatchEvent('resize', rect));

    publisher.start();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (globalThis as any)[$$PublisherInstanceSymbol] = publisher;
};

if (['complete', 'interactive'].includes(document.readyState)) {
    onDOMReady();
} else {
    window.addEventListener('DOMContentLoaded', onDOMReady);
}
