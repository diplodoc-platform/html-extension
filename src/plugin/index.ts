export type {BaseTarget, StylesObject, HTMLRuntimeConfig, EmbeddingMode} from '../types';
export type {PluginOptions} from './transform';
export {getStyles} from './utils';
export {transform} from './transform';
