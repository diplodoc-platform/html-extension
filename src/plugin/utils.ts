import sanitizeHtml from 'sanitize-html';

export const defaultSanitize = (dirtyHtml: string) => sanitizeHtml(dirtyHtml);

export function addHiddenProperty<
    B extends Record<string | symbol, unknown>,
    F extends string | symbol,
    V,
>(box: B, field: F, value: V) {
    if (!(field in box)) {
        Object.defineProperty(box, field, {
            enumerable: false,
            value: value,
        });
    }

    return box as B & {[P in F]: V};
}

export const resizeIframeToFitContent = (iframe: HTMLIFrameElement): void => {
    if (iframe.contentWindow) {
        iframe.height = iframe.contentWindow.document.documentElement.scrollHeight + 'px';
    }
};

export const isHTMLElement = (node: Node): node is HTMLElement =>
    node.nodeType === Node.ELEMENT_NODE;
export const isIFrame = (
    element?: HTMLElement | EventTarget | null,
): element is HTMLIFrameElement =>
    element !== undefined &&
    element instanceof Element &&
    element.tagName.toLowerCase() === 'iframe';

export const setIframeStyles = (block: HTMLIFrameElement, styles: Record<string, string>) => {
    Object.keys(styles).forEach((property) => {
        block.contentWindow?.document.body.style.setProperty(property, styles[property]);
    });
};
