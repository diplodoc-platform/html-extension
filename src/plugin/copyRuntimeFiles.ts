import {dynrequire} from './utils';

const PATH_TO_RUNTIME = '../runtime';

interface CopyRuntimeFilesPaths {
    runtimeJsPath: string;
    output: string;
}

export function copyRuntimeFiles(
    {runtimeJsPath, output}: CopyRuntimeFilesPaths,
    cache: Set<string>,
) {
    const {join, resolve} = dynrequire('node:path');
    const runtimeFiles = {
        'index.js': runtimeJsPath,
    };
    for (const [originFile, outputFile] of Object.entries(runtimeFiles)) {
        const file = join(PATH_TO_RUNTIME, originFile);
        if (!cache.has(file)) {
            cache.add(file);
            copyFile(resolve(__dirname, file), join(output, outputFile));
        }
    }
}

function copyFile(from: string, to: string) {
    const {mkdirSync, copyFileSync} = dynrequire('node:fs');
    const {dirname} = dynrequire('node:path');
    mkdirSync(dirname(to), {recursive: true});
    copyFileSync(from, to);
}
