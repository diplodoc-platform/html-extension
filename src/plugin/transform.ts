import type MarkdownIt from 'markdown-it';

import {directiveParser, registerContainerDirective} from '@diplodoc/directive';

import {ISOLATED_TOKEN_TYPE, SHADOW_TOKEN_TYPE, SRCDOC_TOKEN_TYPE} from '../constants';
import {BaseTarget, EmbeddingMode, Sanitize, SanitizeConfig, StylesObject} from '../types';

import {addHiddenProperty, dynrequire, getStyles} from './utils';
import {copyRuntimeFiles} from './copyRuntimeFiles';
import {makeIsolatedModeEmbedRenderRule} from './renderers/isolated';
import {makeShadowModeEmbedRenderRule} from './renderers/shadow';
import {makeSrcdocModeEmbedRenderRule} from './renderers/srcdoc';

export interface PluginOptions {
    embeddingMode: EmbeddingMode;
    runtimeJsPath: string;
    containerClasses: string;
    bundle: boolean;
    isolatedSandboxHost?: string;
    sanitize?: Sanitize | SanitizeConfig;
    /**
     * @deprecated Use the 'head' method instead.
     */
    styles?: string | StylesObject;
    /**
     * @deprecated Use the 'head' method instead.
     */
    baseTarget?: BaseTarget;
    head?: string;
    sandbox?: boolean | string;
}

type TransformOptions = {
    output?: string;
};

const emptyOptions = {};
const TAG = 'iframe';

const embeddingModeToTokenType = {
    srcdoc: SRCDOC_TOKEN_TYPE,
    shadow: SHADOW_TOKEN_TYPE,
    isolated: ISOLATED_TOKEN_TYPE,
} as const satisfies Record<EmbeddingMode, string>;

const concatStylesIncludeDirectives = (content: string, styles?: string | StylesObject) => {
    if (styles) {
        const stylesContent =
            typeof styles === 'string'
                ? `<link rel="stylesheet" href="${styles}" />`
                : `<style>${getStyles(styles)}</style>`;

        return stylesContent + content;
    }

    return content;
};

type RegisterTransformOptions = Pick<
    PluginOptions,
    'runtimeJsPath' | 'bundle' | 'embeddingMode'
> & {
    output: string;
};

const registerTransform = (
    md: MarkdownIt,
    {embeddingMode, runtimeJsPath, output, bundle}: RegisterTransformOptions,
) => {
    md.use(directiveParser());

    registerContainerDirective(md, {
        name: 'html',
        type: 'code_block',
        container: {
            tag: TAG,
            token: embeddingModeToTokenType[embeddingMode],
        },
        match: (_params, state) => {
            const {env} = state;
            env.meta ||= {};
            env.meta.script ||= [];
            env.meta.style ||= [];
            if (!env.meta.script.includes(runtimeJsPath)) {
                env.meta.script.push(runtimeJsPath);
            }

            addHiddenProperty(env, 'bundled', new Set<string>());

            if (bundle) {
                copyRuntimeFiles({runtimeJsPath, output}, env.bundled);
            }

            return true;
        },
    });
};

export function transform({
    embeddingMode = 'srcdoc',
    runtimeJsPath = '_assets/html-extension.js',
    containerClasses = '',
    bundle = true,
    isolatedSandboxHost,
    sanitize,
    styles,
    baseTarget = '_parent',
    head: headContent,
    sandbox,
}: Partial<PluginOptions> = emptyOptions): MarkdownIt.PluginWithOptions<TransformOptions> {
    const sanitizeHead = typeof sanitize === 'function' ? sanitize : sanitize?.head;
    const sanitizeBody = typeof sanitize === 'function' ? sanitize : sanitize?.body;

    const plugin: MarkdownIt.PluginWithOptions<TransformOptions> = (md, options) => {
        const {output = '.'} = options || {};

        registerTransform(md, {embeddingMode, runtimeJsPath, bundle, output});

        md.renderer.rules[SRCDOC_TOKEN_TYPE] = makeSrcdocModeEmbedRenderRule({
            containerClassNames: containerClasses,
            sandbox,
            embedContentTransformFn: (raw) => {
                const deprecatedHeadContent = concatStylesIncludeDirectives(
                    `<base target="${baseTarget}">` +
                        // It's important to disable script by default in the srcDoc mode
                        `<meta http-equiv="Content-Security-Policy" content="script-src 'none'">`,
                    styles,
                );

                const rawHeadContent = headContent ?? deprecatedHeadContent;
                const headHtml = sanitizeHead?.(rawHeadContent) ?? rawHeadContent;
                const bodyHtml = sanitizeBody?.(raw) ?? raw;

                const head = `<head>${headHtml}</head>`;
                const body = `<body>${bodyHtml}</body>`;
                const html = `<!DOCTYPE html><html>${head}${body}</html>`;

                return html;
            },
        });

        md.renderer.rules[ISOLATED_TOKEN_TYPE] = makeIsolatedModeEmbedRenderRule({
            containerClassNames: containerClasses,
            baseTarget,
            isolatedSandboxHost,
            sandbox,
            embedContentTransformFn: (raw) => concatStylesIncludeDirectives(raw, styles),
        });

        md.renderer.rules[SHADOW_TOKEN_TYPE] = makeShadowModeEmbedRenderRule({
            containerClassNames: containerClasses,
            embedContentTransformFn: (raw) => {
                const withStyles = concatStylesIncludeDirectives(raw, styles);

                // Note that sanitization is only enabled for `shadow` embedding mode,
                // `isolated` mode applies no restrictions â€” beware of vulnerabilities galore!
                // That's why it's really important to host the iframe runtime on an unrelated `dummy` origin.
                return sanitizeBody?.(withStyles) ?? withStyles;
            },
        });
    };

    Object.assign(plugin, {
        collect(input: string, {destRoot}: {destRoot: string}) {
            const MdIt = dynrequire('markdown-it');
            const md = new MdIt().use((md: MarkdownIt) => {
                registerTransform(md, {
                    embeddingMode,
                    runtimeJsPath,
                    bundle,
                    output: destRoot,
                });
            });

            md.parse(input, {});
        },
    });

    return plugin;
}
