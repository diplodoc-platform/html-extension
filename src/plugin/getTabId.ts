import GithubSlugger from 'github-slugger';

import {Tab} from './transform';

const CUSTOM_ID_REGEXP = /\[?{ ?#(\S+) ?}]?/;

const sluggersStorage = new Map<string, GithubSlugger>();

function parseName(name: string) {
    const parts = name.match(CUSTOM_ID_REGEXP);
    if (!parts) {
        return {
            name,
            customAnchor: null,
        };
    }

    return {
        name: name.replace(parts[0], '').trim(),
        customAnchor: parts[1],
    };
}

export function getTabId(tab: Tab, {runId}: {runId: string}) {
    let slugger = sluggersStorage.get(runId);
    if (!slugger) {
        slugger = new GithubSlugger();
        sluggersStorage.set(runId, slugger);
    }

    return slugger.slug(getRawId(tab));
}

export function getTabKey(tab: Tab) {
    return encodeURIComponent(getRawId(tab)).toLocaleLowerCase();
}

export function getName(tab: Tab) {
    return parseName(tab.name).name;
}

function getRawId(tab: Tab): string {
    return parseName(tab.name).customAnchor || tab.name;
}
