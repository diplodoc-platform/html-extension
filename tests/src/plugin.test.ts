import MarkdownIt from 'markdown-it';
import transform from '@diplodoc/transform';
import dd from 'ts-dedent';

import {type PluginOptions, transform as htmlTransform} from '../../src/plugin';

const html = (text: string, opts?: Partial<PluginOptions>) => {
    return transform(text, {
        needToSanitizeHtml: false,
        plugins: [htmlTransform({bundle: false, ...opts})],
    }).result.html;
};

const tokens = (text: string, opts?: Partial<PluginOptions>) => {
    const md = new MarkdownIt().use(htmlTransform({bundle: false, ...opts}));
    return md.parse(text, {});
};

describe('HTML extension â€“ plugin', () => {
    it('should render html block', () => {
        expect(
            html(
                dd`
            :::html
            <div class="html-div">content</div>
            :::
            `,
                {embeddingMode: 'srcdoc'},
            ),
        ).toMatchSnapshot();
    });

    it('should generate html token', () => {
        expect(
            tokens(
                dd`


            :::html
            <div class="html-div">content</div>
            :::
            `,
                {embeddingMode: 'srcdoc'},
            ),
        ).toMatchSnapshot();
    });

    it('should render HTML in sandboxed iframe', () => {
        expect(
            html(
                dd`
            :::html
            <div class="html-div">content</div>
            :::
            `,
                {
                    embeddingMode: 'srcdoc',
                    sandbox:
                        'allow-popups allow-popups-to-escape-sandbox allow-top-navigation-by-user-activation allow-same-origin',
                },
            ),
        ).toMatchSnapshot();
    });

    it('should apply sanitize function', () => {
        expect(
            html(
                dd`
            :::html
            <p>should-be-replaced</p>
            :::
            `,
                {
                    embeddingMode: 'srcdoc',
                    sanitize: (html) => html.replace('should-be-replaced', 'sanitized'),
                },
            ),
        ).toMatchSnapshot();
    });

    it('should apply sanitize head and body functions', () => {
        expect(
            html(
                dd`
            :::html
            <p>replace-head</p>
            <p>replace-body</p>
            :::
            `,
                {
                    embeddingMode: 'srcdoc',
                    head: '<title>replace-head | replace-body</title>',
                    sanitize: {
                        head: (html) => html.replace('replace-head', 'sanitized'),
                        body: (html) => html.replace('replace-body', 'sanitized'),
                    },
                },
            ),
        ).toMatchSnapshot();
    });
});
