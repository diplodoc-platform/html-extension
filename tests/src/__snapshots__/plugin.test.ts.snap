// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML extension – plugin should apply sanitize function 1`] = `
<iframe srcdoc="&lt;!DOCTYPE html&gt;&lt;html&gt;&lt;head&gt;&lt;base target=&quot;_parent&quot;&gt;&lt;meta http-equiv=&quot;Content-Security-Policy&quot; content=&quot;script-src 'none'&quot;&gt;&lt;/head&gt;&lt;body&gt;&lt;p&gt;sanitized&lt;/p&gt;
&lt;/body&gt;&lt;/html&gt;"
        frameborder="0"
        style="width:100%"
        data-yfm-sandbox-mode="srcdoc"
>
</iframe>
`;

exports[`HTML extension – plugin should apply sanitize head and body functions 1`] = `
<iframe srcdoc="&lt;!DOCTYPE html&gt;&lt;html&gt;&lt;head&gt;&lt;title&gt;sanitized | replace-body&lt;/title&gt;&lt;/head&gt;&lt;body&gt;&lt;p&gt;replace-head&lt;/p&gt;
&lt;p&gt;sanitized&lt;/p&gt;
&lt;/body&gt;&lt;/html&gt;"
        frameborder="0"
        style="width:100%"
        data-yfm-sandbox-mode="srcdoc"
>
</iframe>
`;

exports[`HTML extension – plugin should generate html token 1`] = `
[
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": <div class="html-div">
  content
</div>,
    "hidden": false,
    "info": "html",
    "level": 0,
    "map": [
      2,
      5,
    ],
    "markup": ":::",
    "meta": null,
    "nesting": 0,
    "tag": "iframe",
    "type": "yfm_html_block",
  },
]
`;

exports[`HTML extension – plugin should render HTML in sandboxed iframe 1`] = `
<iframe srcdoc="&lt;!DOCTYPE html&gt;&lt;html&gt;&lt;head&gt;&lt;base target=&quot;_parent&quot;&gt;&lt;meta http-equiv=&quot;Content-Security-Policy&quot; content=&quot;script-src 'none'&quot;&gt;&lt;/head&gt;&lt;body&gt;&lt;div class=&quot;html-div&quot;&gt;content&lt;/div&gt;
&lt;/body&gt;&lt;/html&gt;"
        frameborder="0"
        style="width:100%"
        data-yfm-sandbox-mode="srcdoc"
        sandbox="allow-popups allow-popups-to-escape-sandbox allow-top-navigation-by-user-activation allow-same-origin"
>
</iframe>
`;

exports[`HTML extension – plugin should render html block 1`] = `
<iframe srcdoc="&lt;!DOCTYPE html&gt;&lt;html&gt;&lt;head&gt;&lt;base target=&quot;_parent&quot;&gt;&lt;meta http-equiv=&quot;Content-Security-Policy&quot; content=&quot;script-src 'none'&quot;&gt;&lt;/head&gt;&lt;body&gt;&lt;div class=&quot;html-div&quot;&gt;content&lt;/div&gt;
&lt;/body&gt;&lt;/html&gt;"
        frameborder="0"
        style="width:100%"
        data-yfm-sandbox-mode="srcdoc"
>
</iframe>
`;
